### ISSUE TITLE
Null for optional boolean field on complex input value throws GraphqlException

### ISSUE THREAD

**Issue by derpfeiffer** [andimarek/graphql-java#64]
In the latest commits there has been a code change here:

graphql.execution.ValuesResolver#coerceValueForInputObjectField

It used to return null when the input was null. 

Suppose you have a complex input type object having a boolean field. The boolean field is optional and has a default value of false in the schema.
Sending a mutation with this complex input type object (as required by relay) always results in a exception. The coerce code does not accept any of the primitives to be null at this point (Long, Boolean,...)

When first starting to parse, it parses the main arguments in the mutation request with this logic:

if (inputValue == null && variableDefinition.getDefaultValue() != null) {
return coerceValueAst(type, variableDefinition.getDefaultValue(), null);
}
Which is present here: graphql.execution.ValuesResolver#getVariableValue

However when parsing input object fields other logic is used an no optional values are allowed. Is there any reason for this ? Seems to me the same logic applies there and it should take into account optional fields with default values.

**Comment by andimarek** [comment]
Hi,

could you please create a small example of what you mean? Or maybe past the stacktrace of the exception you mentioned? 

This would be really helpful.

Thanks!

**Comment by derpfeiffer** [comment]
Here is the stacktrace:

```java
graphql.GraphQLException
	at graphql.Scalars$5.serialize(Scalars.java:121)
	at graphql.Scalars$5.parseValue(Scalars.java:127)
	at graphql.execution.ValuesResolver.coerceValueForScalar(ValuesResolver.java:92)
	at graphql.execution.ValuesResolver.coerceValue(ValuesResolver.java:68)
	at graphql.execution.ValuesResolver.coerceValueForInputObjectField(ValuesResolver.java:84)
	at graphql.execution.ValuesResolver.coerceValue(ValuesResolver.java:74)
	at graphql.execution.ValuesResolver.coerceValue(ValuesResolver.java:76)
	at graphql.execution.ValuesResolver.coerceValueForInputObjectField(ValuesResolver.java:84)
	at graphql.execution.ValuesResolver.coerceValue(ValuesResolver.java:74)
	at graphql.execution.ValuesResolver.coerceValue(ValuesResolver.java:76)
	at graphql.execution.ValuesResolver.getVariableValue(ValuesResolver.java:59)
	at graphql.execution.ValuesResolver.getVariableValues(ValuesResolver.java:16)
	at graphql.execution.ExecutionContextBuilder.build(ExecutionContextBuilder.java:55)
	at graphql.execution.Execution.execute(Execution.java:32)
	at graphql.GraphQL.execute(GraphQL.java:78)
	at graphql.GraphQL.execute(GraphQL.java:55)
```

Suppose you have a mutation input object with the following schema:
```javascript
{
    "kind": "INPUT_OBJECT",
    "name": "AddFriend",
    "description": "The friend mutation input",
    "fields": null,
    "inputFields": [
      {
        "name": "name",
        "description": "The friends name",
        "type": {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null
        },
        "defaultValue": "Jef"
      },
      {
        "name": "isFriend",
        "description": "True if friend, false otherwise",
        "type": {
          "kind": "SCALAR",
          "name": "Boolean",
          "ofType": null
        },
        "defaultValue": "false"
      }
    ],
    "interfaces": null,
    "enumValues": null,
    "possibleTypes": null
}
```

It has one field name and one field isFriend which is of type boolean. This boolean is optional and has a default value.

However when you do a mutation request like this:

```javascript
mutation myFriendMutation($input: AddFriend) {
  addFriend(input: $input) {
    clientMutationId
  }
}
```
And no boolean field in the input object, if e.g. these are the variables:
```javascript
{
  "input": {
    "name": "Emma"
  }
}
```
The above will always result in the stacktrace. A string can be optional, as the coerce method checks for null, however this is not the case for booleans and I think it will also fail for long values.

Basically the problem is also if you nest fields inside a input object. The code path when you would do a request without an input object and with a flat set of arguments, will work as that takes the null values into account and performs different logic.  'coerceValueAst(type, variableDefinition.getDefaultValue(), null);
}' like I mentioned above.

I don't see a reason why a field in a complex input object should be treated differently than a field on the top level passed in as a argument ?

**Comment by andimarek** [comment]
Thanks for the explanation ... this makes it clear. After having a quick look I think the  problem is the serialize function of boolean: It should return null and not throwing exceptions. And then 
this code: https://github.com/andimarek/graphql-java/blob/master/src/main/java/graphql/execution/ValuesResolver.java#L85 should use the default value.

**Comment by derpfeiffer** [comment]
Indeed. That would fix it.
But also looking at all other types in scalars (Int, Long, Float, ID) they will all suffer from the same problem. They can currently never be optional in a Input object

**Comment by andimarek** [comment]
You are right: This bug affected other scalars too. It should be fixed with this commit: https://github.com/andimarek/graphql-java/commit/43d5f57abe49a16349676aa9c144ce1fa9670187
